### ISSUE TITLE
workspace <n> output <direction> can crash i3

### ISSUE THREAD

**Issue by orestisf1993** [i3/i3#3208]
Output of `i3 --moreversion 2>&- || i3 --version`:
```
Binary i3 version:  4.15-71-g45be56be (2018-03-27, branch "next") © 2009 Michael Stapelberg and contributors
```

URL to a logfile as per https://i3wm.org/docs/debugging.html:

https://logs.i3wm.org/logs/5678957424803840.bz2

**What I did:**
```
# i3 config file (v4)

workspace 5 output left
# I guess it should work with real outputs as well
fake-outputs 800x500+0+0,800x500+800+0,800x500+800+500,800x500+0+500
```
Run `i3-msg workspace 2, rename workspace 1 to 5`

**What I saw:**
```
==5487==ERROR: AddressSanitizer: heap-use-after-free on address 0x614000001e9c at pc 0x55556eb52de7 bp 0x7ffe742200c0 sp 0x7ffe742200b0
READ of size 4 at 0x614000001e9c thread T0
    #0 0x55556eb52de6 in con_get_workspace ../../i3/src/con.c:421
    #1 0x55556eb4380a in cmd_rename_workspace ../../i3/src/commands.c:2012
    #2 0x55556eb46957 in GENERATED_call parser/GENERATED_command_call.h:22
    #3 0x55556eb4a3aa in next_state ../../i3/src/commands_parser.c:185
    #4 0x55556eb4c4c4 in parse_command ../../i3/src/commands_parser.c:346
    #5 0x55556eba0ebf in handle_run_command ../../i3/src/ipc.c:124
    #6 0x55556ebac906 in ipc_receive_message ../../i3/src/ipc.c:1248
    #7 0x7f73d9fa44e2 in ev_invoke_pending (/usr/lib/libev.so.4+0x44e2)
    #8 0x7f73d9fa7ad5 in ev_run (/usr/lib/libev.so.4+0x7ad5)
    #9 0x55556ebb7a3f in ev_loop /usr/local/include/ev.h:835
    #10 0x55556ebc3495 in main ../../i3/src/main.c:964
    #11 0x7f73d8978f49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)
    #12 0x55556eb01749 in _start (/home/orestis/Documents/programming/i3/build/i3+0x189749)

0x614000001e9c is located 92 bytes inside of 448-byte region [0x614000001e40,0x614000002000)
freed by thread T0 here:
    #0 0x7f73dcef9711 in __interceptor_free /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:45
    #1 0x55556eb4f19c in con_free ../../i3/src/con.c:96
    #2 0x55556ebfffec in tree_close_internal ../../i3/src/tree.c:340
    #3 0x55556eb5ee2d in con_on_remove_child ../../i3/src/con.c:1986
    #4 0x55556ec14a63 in workspace_move_to_output ../../i3/src/workspace.c:1004
    #5 0x55556eb437ed in cmd_rename_workspace ../../i3/src/commands.c:2009
    #6 0x55556eb46957 in GENERATED_call parser/GENERATED_command_call.h:22
    #7 0x55556eb4a3aa in next_state ../../i3/src/commands_parser.c:185
    #8 0x55556eb4c4c4 in parse_command ../../i3/src/commands_parser.c:346
    #9 0x55556eba0ebf in handle_run_command ../../i3/src/ipc.c:124
    #10 0x55556ebac906 in ipc_receive_message ../../i3/src/ipc.c:1248
    #11 0x7f73d9fa44e2 in ev_invoke_pending (/usr/lib/libev.so.4+0x44e2)

previously allocated by thread T0 here:
    #0 0x7f73dcef9ce1 in __interceptor_calloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:70
    #1 0x55556ec378c1 in scalloc ../../i3/libi3/safewrappers.c:31
    #2 0x55556eb4dc49 in con_new_skeleton ../../i3/src/con.c:40
    #3 0x55556eb4e5e2 in con_new ../../i3/src/con.c:71
    #4 0x55556ec0c726 in create_workspace_on_output ../../i3/src/workspace.c:178
    #5 0x55556ebdc34d in init_ws_for_output ../../i3/src/randr.c:520
    #6 0x55556eb82a0e in fake_outputs_init ../../i3/src/fake_outputs.c:77
    #7 0x55556ebc02b1 in main ../../i3/src/main.c:714
    #8 0x7f73d8978f49 in __libc_start_main (/usr/lib/libc.so.6+0x20f49)
```

**What I expected instead:**
I guess we shouldn't interpret directional outputs as such in `workspace <workspace> output <output>`.

**Comment by Airblader** [comment]
> I guess we shouldn't interpret directional outputs as such in workspace <workspace> output <output>.

Agreed. :-) I mean, we allow it, but it should be interpreted as "the output named »left«".

**Comment by ograff** [comment]
Ultimately, I'll be submitting the change I make as part of a school project (https://web.eecs.umich.edu/~weimerw/481/hw6.html). As part of that project I have to come up with a proposed timeline for getting this work done. I figure this info might be helpful for anyone waiting on a fix for this defect, so I've posted that here as well.

### Time Plan

By 3/30: Build and run i3 successfully as is in a VM. Run i3 test suite. Recreate the bug locally.

By 4/2: Have a reasonable idea of the root cause of the issue.

By 4/4: Implement a fix locally for the bug. Run a cursory test.

By 4/6: Implement a test case that checks that the rename command works as expected
based on the failure reported in the original issue. Verify that the test
case would have failed without the fix I plan to implement.

By 4/7: Review other PRs on the i3 project to see common mistakes,
verify I have fixed those on my contribution.

By 4/8: Implement any further test cases needed. Submit PR to original project
for comment / rework.

Complete rework as comments are made by regular contributors / reviewers.

(Ideally) By 4/16: Changes merged into i3 `next` branch.

**Comment by orestisf1993** [comment]
Some comments:

1. If you run linux on your machine you can skip the VM and run the locally built i3 inside Xephyr. Very convenient since it allows you to crash i3 without affecting the rest of your system.
2. The stack trace above could be a bit misleading.

@Airblader 
> Agreed. :-) I mean, we allow it, but it should be interpreted as "the output named »left«".

Pedantic here but an output named "left" would cause a variety of problems since it will be interpreted as a directional output first.

**Comment by ograff** [comment]
@orestisf1993 
1. I wasn't familiar with Xephyr, thanks!

2. I got the impression that the root cause was probably not at the time of the stack trace

3. Thanks, I'm sure I'll have questions 😃 

With regards to interpreting "left" not as a directional output in `workspace <workspace> output <output>`, but having an output named "left" being problematic, I assume the expected behavior should be:
"Whatever happens when `<output>` names a nonexistent output?"

**Comment by Airblader** [comment]
@orestisf1993 My point is exactly that the "parse a directional output" should be skipped. Which means if the user had a output named "left", it'd assign it to that one. Since that's very unlikely, the assignment would be skipped since the output doesn't exist.

**Comment by orestisf1993** [comment]
Yes, I understand, but this could become a bit tricky / complicated to fix. We have to constantly make sure that no output used in an assignment directive is passed in `get_output_from_string`. If we abort directional output names (**not** non-existing outputs, they could appear later) when the config file is loaded it could lead to a cleaner solution:

1. Still allow "primary" to be used -- shouldn't this be handled by `get_output_from_string`?
2. The rest of the code can assert non-directional `Workspace_Assignment.output`.
3. The error will be logged early, instead of silently failing.

**Comment by Airblader** [comment]
I haven't looked at the implementation, but since we have to allow specifying output names (which is the entire point of the command), we have to allow any kind of word for the argument. That means a check on config parsing level would have to explicitly look for these keywords and then error out; I don't see a good reason to do so. I'd just make sure that we don't try to parse the special keywords for these assignments, but instead only parse primary + named outputs.

**Comment by ograff** [comment]
My understanding so far, is that this breaks because when the code tries to rename workspace 1 to workspace 5, it moves it to the output left of workspace 1, which is the output workspace 2 is on. Workspace 2 now becomes orphaned, so when we try to show it again, after renaming 1 to 5, its been removed and freed because it no longer had an output.

----

Based on the read through I've done so far, it seems like both config validation, and treating outputs specified in `Workspace_Assignment.output` as primary + named outputs should be fairly straightforward. 

There is something nice about rejecting these configs as they don't make much sense, but I'm actually partial to parsing them as primary + named output. Detecting keywords in the config could get troublesome, for example, if new keywords were to be added for outputs, the assertions on `Workspace_Assignment.output` would need to be updated.

Treating the output name as non-directional if it comes from `Workspace_Assignment.output` would result in falling back to the default behavior if a workspace is assigned to a named output that doesn't exist, which as appears to be a nice error log message, and no actual movement of the workspace, therefore no freeing of the original workspace.

----

With regards to `primary`, is there a chance the same issue could occur? Lets say the config is

```
# i3 config file (v4)

workspace 5 output primary
# I guess it should work with real outputs as well
fake-outputs 800x500+0+0,800x500+800+0,800x500+800+500,800x500+0+500
```

couldn't

`i3-msg focus output primary, rename workspace 1 to 5` cause a similar problem.

So far I've been unable to actually create that issue, probably because I can't get randr to report multiple monitors with one as primary using my Xephyr setup, but I was wondering if it was a possible problem, or if `i3` would start with workspace 5 displayed on the primary output based on that config, making it a non issue.

**Comment by orestisf1993** [comment]
`fake-outputs` supports faking the primary output as well.

Try

`fake-outputs 800x500+0+0P,800x500+800+0,800x500+800+500,800x500+0+500`

verify that it works with `focus output primary`

**Comment by ograff** [comment]
It indeed appears that a similar issue occurs withe the following setup:

config:

```
# i3 config file (v4)

workspace 5 output primary
fake-outputs 800x500+0+0P,800x500+800+0,800x500+800+500,800x500+0+500
```

Ran `i3-msg focus output primary, rename workspace 2 to 5`

`i3` then crashed with the same / similar stack traces as posted in the original issue, but I have pasted them here as well.
```
==2757==ERROR: AddressSanitizer: heap-use-after-free on address 0x61400000149c at pc 0x5598567b89b1 bp 0x7ffe8a592240 sp 0x7ffe8a592230
READ of size 4 at 0x61400000149c thread T0
    #0 0x5598567b89b0 in con_get_workspace ../../i3/src/con.c:421
    #1 0x5598567aee06 in cmd_rename_workspace ../../i3/src/commands.c:1996
    #2 0x5598567b0983 in GENERATED_call parser/GENERATED_command_call.h:22
    #3 0x5598567b34a8 in next_state ../../i3/src/commands_parser.c:185
    #4 0x5598567b48cf in parse_command ../../i3/src/commands_parser.c:346
    #5 0x5598567ef94f in handle_run_command ../../i3/src/ipc.c:124
    #6 0x5598567f7668 in ipc_receive_message ../../i3/src/ipc.c:1248
    #7 0x7f1676eafd72 in ev_invoke_pending (/usr/lib/x86_64-linux-gnu/libev.so.4+0x3d72)
    #8 0x7f1676eb33dd in ev_run (/usr/lib/x86_64-linux-gnu/libev.so.4+0x73dd)
    #9 0x5598567ff0c2 in ev_loop /usr/include/ev.h:835
    #10 0x559856807dc1 in main ../../i3/src/main.c:964
    #11 0x7f16765781c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)
    #12 0x559856787c99 in _start (/usr/local/bin/i3+0x4dc99)

0x61400000149c is located 92 bytes inside of 448-byte region [0x614000001440,0x614000001600)
freed by thread T0 here:
    #0 0x7f1679db07b8 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xde7b8)
    #1 0x5598567b65ab in con_free ../../i3/src/con.c:96
    #2 0x559856831e97 in tree_close_internal ../../i3/src/tree.c:340
    #3 0x5598567c0e2b in con_on_remove_child ../../i3/src/con.c:1993
    #4 0x559856840075 in workspace_move_to_output ../../i3/src/workspace.c:1014
    #5 0x5598567aeded in cmd_rename_workspace ../../i3/src/commands.c:1993
    #6 0x5598567b0983 in GENERATED_call parser/GENERATED_command_call.h:22
    #7 0x5598567b34a8 in next_state ../../i3/src/commands_parser.c:185
    #8 0x5598567b48cf in parse_command ../../i3/src/commands_parser.c:346
    #9 0x5598567ef94f in handle_run_command ../../i3/src/ipc.c:124
    #10 0x5598567f7668 in ipc_receive_message ../../i3/src/ipc.c:1248
    #11 0x7f1676eafd72 in ev_invoke_pending (/usr/lib/x86_64-linux-gnu/libev.so.4+0x3d72)

previously allocated by thread T0 here:
    #0 0x7f1679db0d38 in __interceptor_calloc (/usr/lib/x86_64-linux-gnu/libasan.so.4+0xded38)
    #1 0x55985685a28d in scalloc ../../i3/libi3/safewrappers.c:31
    #2 0x5598567b58b4 in con_new_skeleton ../../i3/src/con.c:40
    #3 0x5598567b5efe in con_new ../../i3/src/con.c:71
    #4 0x55985683ae32 in create_workspace_on_output ../../i3/src/workspace.c:202
    #5 0x55985681876a in init_ws_for_output ../../i3/src/randr.c:516
    #6 0x5598567dab23 in fake_outputs_init ../../i3/src/fake_outputs.c:77
    #7 0x559856805ad5 in main ../../i3/src/main.c:714
    #8 0x7f16765781c0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x211c0)

```

The log for this particular way of causing this crash is here: https://logs.i3wm.org/logs/5733572765810688.bz2

----

Given that this is still an issue with primary in addition to the other directions, do you guys think we should treat primary as any of the other directions for the sake of renaming (ie treat it as a proper name)? Is there a deeper issue we should be addressing? It seems like perhaps, this issue would occur if:

There was an output named `foo` and a config line
`workspace 5 output foo`

If output `foo` is displaying workspace 2, and then someone runs `i3-msg rename workspace 1 to 5` this crash would still occur. Any tips on creating named outputs using `fake-outputs`? (I plan to go read the code as well to try and find out)

**Comment by ograff** [comment]
Interesting, this does not occur when using a named output (`fake-1`) for workspace 5. Instead, output `fake-1`'s initially created workspace is workspace 5, therefore attempting to rename workspace 1 to 5 fails because workspace 5 already exists. 

You can still force this crash to occur using named outputs with the following sequence of steps:

```
# i3 config file (v4)

workspace 5 output fake-1
fake-outputs 800x500+0+0P,800x500+800+0,800x500+800+500,800x500+0+500
```

Launch i3

Run `i3-msg rename workspace 5 to 6, workspace 6, rename workspace 1 to 5`

The result is once again a crash with a similar stack trace, I'm going to avoid spamming the issue with additional logs and stack traces that say much the same thing unless anyone is particularly interested in them.

----

I'm not 100% sure what the most correct fix for this is, but I think it probably is bigger than just treating directional outputs as named outputs. I think the workspace being renamed should be added to the content container of the appropriate output, instead of replacing (and freeing) the current workspace on that output. Perhaps this issue should cover just changing how output names are handled, and a different issue should be opened to address the deeper potential crash during rename issue.

Thoughts?

**Comment by Airblader** [comment]
I haven't had time to go through the details here but splitting it in the way you suggest sounds good to me! The "primary" output definitely needs to be allowed for assignments still.

**Comment by orestisf1993** [comment]
To fix this issue (but not `i3-msg rename workspace 5 to 6, workspace 6, rename workspace 1 to 5`) you can try to validate the name in cmd_rename using `Output *get_output_by_name(const char *name, const bool require_active)`.

**Comment by ograff** [comment]
@orestisf1993 Sounds good. I'm going to tackle the directional issue first. My plan is to add parameters to `move_workspace_to_output` and `get_output_from_string` to specify if directions should be treated as literals or keywords (with `primary` always being treated as a keyword). Does this seem like a reasonable solution?

**Comment by orestisf1993** [comment]
*`workspace_move_to_output` right? Well the new parameter is only going to be used once right? You could put the new code in `cmd_rename_workspace` to minimize the changes. For reference the function is only called 2 times:

```
src/commands.c
1131:        bool success = workspace_move_to_output(ws, name);
1993:        workspace_move_to_output(workspace, assignment->output);
```

Subjective/code style related issue, I'll leave the final call to @Airblader.

**Comment by Airblader** [comment]
Instead of a new parameter, it might make sense to make the function take the (already parsed) output as a parameter instead. Then parsing the string into the correct output is done outside the function giving the caller control over how this is done.

Haven't looked at the code, though. I'm a bit short on time and will be until Monday at least. I do trust that @orestisf1993 can make a good decision here; it's really not all that crucial anyway :-)

**Comment by ograff** [comment]
I like the solution suggested by @Airblader 
I'll have a PR up with at least the implementation changes, and hopefully the appropriate tests by sometime tomorrow.

**Comment by ograff** [comment]
PR is up at #3221
